File: 0286-walls-and-gates/0286-walls-and-gates.java
class Solution {
    static int[][] dirs = { { -1, 0 }, { 0, 1 }, { 1, 0 }, { 0, -1 } };
    final int INF = Integer.MAX_VALUE;

    public void wallsAndGates(int[][] rooms) {
        int n = rooms.length;
        int m = rooms[0].length;
        Queue<int[]> q = new ArrayDeque<>();
        boolean[][] vis = new boolean[n][m];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (rooms[i][j] == 0) {
                    vis[i][j] = true;
                    q.offer(new int[] { 0, i, j });
                }
            }
        }

        while (!q.isEmpty()) {
            int t = q.peek()[0];
            int r = q.peek()[1];
            int c = q.peek()[2];
            q.poll();

            rooms[r][c] = Math.min(rooms[r][c], t);
            for (int[] dir : dirs) {
                int nr = r + dir[0];
                int nc = c + dir[1];

                if (nr >= 0 && nr < n && nc >= 0 && nc < m && !vis[nr][nc] && rooms[nr][nc] == INF) {
                    q.offer(new int[] { t + 1, nr, nc });
                    vis[nr][nc] = true;
                }
            }
        }
    }
}

File: 0286-walls-and-gates/0286-walls-and-gates.cs
public class Solution {
    static int[, ] dirs = {{-1, 0}, {0, 1}, {1, 0}, {0 ,-1}};
    public void WallsAndGates(int[][] rooms) {
        int n = rooms.Length;
        int m = rooms[0].Length;
        int INF = Int32.MaxValue;

        var q = new Queue<(int, int, int)>();
        var vis = new bool[n, m];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (rooms[i][j] == 0) {
                    vis[i, j] = true;
                    q.Enqueue((0, i, j));
                }
            }
        }

        //Main logic starts
        while (q.Count > 0) {
            var (t, r, c) = q.Dequeue();
            rooms[r][c] = Math.Min(rooms[r][c], t);

            for (int i = 0; i < 4; i++) {
                int nr = r + dirs[i, 0];
                int nc = c + dirs[i, 1];

                if (nr >= 0 && nr < n && nc >= 0 && nc < m && !vis[nr, nc] && rooms[nr][nc] == INF) {
                    vis[nr, nc] = true;
                    q.Enqueue((t + 1, nr, nc));
                }
            }
        }
    }
}